### DISCUSSION TITLE
How to use Pydantic validator for a class based dependency

### DISCUSSION THREAD

**Discussion by alukach** [fastapi/fastapi#3426] [Questions] [answered]
I feel like I'm missing something basic from the docs, but what is the recommended way to validate a [class-based dependency](https://fastapi.tiangolo.com/tutorial/dependencies/classes-as-dependencies/) so that a `400` error is returned to the user?

Here's a simple example app:

```py
from datetime import date
from typing import Optional

from fastapi import FastAPI, Depends
from pydantic import BaseModel, root_validator

app = FastAPI()


class DateRange(BaseModel):
    since: Optional[date] = date(2010, 1, 1)
    to: Optional[date] = date(2020, 1, 1)

    @root_validator
    def assert_since_gt_to(cls, values):
        if values.get("since") >= values.get("to"):
            raise ValueError("'since' argument must be less than 'to' argument")
        return values


@app.get("/")
def echo(daterange: DateRange = Depends()):
    return daterange
```

An example usage:

```sh
$ curl 'http://localhost:8000/'
{"since":"2010-01-01","to":"2020-01-01"}
```

But when I try something where validation fails, I get an `Internal Server Error`.

```sh
$ curl 'http://localhost:8000/?since=2020-01-01&to=2020-01-01'
Internal Server Error
```

I have read through [`RequestValidationError` vs `ValidationError`](https://fastapi.tiangolo.com/tutorial/handling-errors/#requestvalidationerror-vs-validationerror) where it distinguishes how `ValidationErrors` are intended to return `500` as they indicate a response failure, but I'm not understanding how I should/could recast those errors as `RequestValidationErrors`...

---

Related: https://github.com/tiangolo/fastapi/issues/3761

**Comment by alukach** [comment, marked as answer]
After reading through https://github.com/tiangolo/fastapi/issues/318, I believe this is the best that I could come up with:

```py
@dataclass
class DateRange:
    since: Optional[date] = date(2010, 1, 1)
    to: Optional[date] = date(2020, 1, 1)

    def __post_init__(self):
        if self.since >= self.to:
            raise HTTPException(
                422,
                detail=[
                    {
                        "loc": ["query", loc],
                        "msg": "'since' argument must be less than 'to' argument",
                        "type": "value_error.date",
                    }
                    for loc in ["to", "since"]
                ],
            )
```

I don't love this solution, I feel that manually constructing the error message is non-ideal and introduces complexity to the developer (it takes work to write error messages that match FastAPI's validation error format), however I can live with this.

**Reply by ghandic** [reply]
Similar solution but you can also use pydantic `BaseModel` and their `@validator` for example

```python
class CreateUser(BaseModel):
    username: str
    password: str

    @validator("username")
    def username_in_db(cls, v):
        if v not in db:
            raise HTTPException(status_code=403, detail="Username not found")
        return v
```

Feel free to mark your discussion as answered, unless you would have some suggestions for how you would like it to be handled by FastAPI?

I see your issue though, you're coupling FastAPI and pydantic model together - ideally, you could handle the `ValidationError` raised by pydantic and intercept the exception - this is possible and probably the go-to method for larger applications when working with multiple team members

**Reply by alukach** [reply]
@ghandic Thanks for your reply.  I admit that I don't recall _why_ I switched from using `BaseModel` to `dataclass`, but I feel like there was a reason.  I think that only added confusing details to the actual question here and you're correct, I could stick with the `BaseModel` as you demonstrated.

Nonetheless, my question is not fully answered to my liking (largely due to the my having poorly described my original question).  **What I'm truly trying to get is for a way to raise validation exceptions in class-based dependencies that would render a response similar to how Pydantic/FastAPI renders validation errors for non-class based parameters.**

You've illustrated how I can throw an `HTTPException`, however that does _not_ return error object that conforms to other FastAPI validation errors (i.e. it does not have a `loc`, `msg`, and `type` property).  I had to manually craft such an exception in my example.

Let's take a look at this example code:

```py
from datetime import date
import fastapi
import pydantic

app = fastapi.FastAPI()


class DateRangeError(pydantic.PydanticValueError):
    msg_template = "'since' argument must be less than 'to' argument"


class DateRange(pydantic.BaseModel):
    since: date
    to: date

    @pydantic.root_validator
    def assert_since_gt_to(cls, values):
        if not all(values.get(f) for f in ("since", "to")):
            return
        if values["since"] >= values["to"]:
            raise DateRangeError()


@app.get("/body")
def echo(daterange: DateRange):
    return daterange


@app.get("/params")
def echo(daterange: DateRange = fastapi.Depends()):
    return daterange
```


I've since learned that throwing a `PydanticValueError` exception allows FastAPI to return a correctly-formatted `422` response when analyzing the _body_ of a request:

```sh
$ curl -X GET http://127.0.0.1:8000/body \
    --data '{"since": "2020-01-01", "to": "1998-01-01"}' \
    -H 'Content-Type: application/json'
{"detail":[{"loc":["body","__root__"],"msg":"'since' argument must be less than 'to' argument","type":"value_error.daterange"}]}
```

However, when analyzing the query parameters of a request via a class-based dependency, FastAPI does not handle `PydanticValueError` exceptions:

```sh
$ curl -X GET http://127.0.0.1:8000/params?since=2020-01-01&to=1998-01-01
Internal Server Error
```

```py
Traceback (most recent call last):
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/uvicorn/protocols/http/httptools_impl.py", line 398, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/fastapi/applications.py", line 199, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/starlette/routing.py", line 580, in __call__
    await route.handle(scope, receive, send)
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/starlette/routing.py", line 241, in handle
    await self.app(scope, receive, send)
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/starlette/routing.py", line 52, in app
    response = await func(request)
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/fastapi/routing.py", line 209, in app
    solved_result = await solve_dependencies(
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/fastapi/dependencies/utils.py", line 552, in solve_dependencies
    solved = await run_in_threadpool(call, **sub_values)
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/starlette/concurrency.py", line 40, in run_in_threadpool
    return await loop.run_in_executor(None, func, *args)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/alukach/.local/share/virtualenvs/pydantic-example/lib/python3.8/site-packages/pydantic/main.py", line 406, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for DateRange
__root__
  'since' argument must be less than 'to' argument (type=value_error.daterange)
```

What I would expect is an error like the following, which is similar to any other validation error (e.g. if I passed in a non-parsable value for a date, like `from=foo`):

```json
{"detail":[{"loc":["query"],"msg":"'since' argument must be less than 'to' argument","type":"value_error.daterange"}]}
```

Does this make sense?  I am starting to feel like this is a bug/feature-request rather than a discussion, however I am still hoping that there is a supported way to get this behavior in FastAPI today that doesn't require me to manually construct `HTTPException` errors by hand (that feels not in accordance to the zen of FastAPI).

**Reply by ghandic** [reply]
I'll look into this today 👍🏻

**Reply by ghandic** [reply]
https://github.com/tiangolo/fastapi/pull/3953 I have created a basic PR to show where to intercept and address the issue. Feel free to mark your discussion as answered and then we can continue the discussion of implementation and side effects on the PR?

**Reply by alukach** [reply]
Amazing, thank you for looking into this!
